WalletWasabi.Tests/UnitTests/Crypto/ZeroKnowledge/LinearRelationTests.cs

```diff
@@ -75,35 +76,114 @@ public void IgnoredWitnessComponents()
 			// same that sub-response should be the same for the same nonce
 			Assert.False(response1 == response2);
 			Assert.True(response1.First() == response2.First());
 		}
+
+		[Fact]
+		public void StatementAndKnowledge()
+		{
+			var x = new Scalar(42);
+			var a = x * Generators.Gg;
+			var b = x * Generators.Gh;
+
+			// Discrete log equality (Chaum-Pedersen proof)
+			var statement = new Statement(
+				new Equation(a, new GroupElementVector(Generators.Gg)),
+				new Equation(b, new GroupElementVector(Generators.Gh))
+			);
+
+			// Check that simulation produces valid transcripts
+			var challenge = new Scalar(13);
+			var simulatedResponses = new[]
+			{
+				new ScalarVector(Scalar.One),
+				new ScalarVector(Scalar.One),
+			};
+
+			var simulatedPublicNonces = statement.SimulatePublicNonces(challenge, simulatedResponses);
+			Assert.True(statement.CheckVerificationEquation(simulatedPublicNonces, challenge, simulatedResponses));
+
+			// Create transcripts using a witness to the relation
+			var knowledge = new Knowledge(statement, new ScalarVector(x));
+			var allSecretNonces = new[] { new ScalarVector(new Scalar(7)), new ScalarVector(new Scalar(11)) };
+			var publicNonces = new GroupElementVector(statement.Equations.Zip(allSecretNonces, (equation, secretNonces) => secretNonces * equation.Generators));
+			var allResponses = knowledge.RespondToChallenge(challenge, allSecretNonces);
+			Assert.True(statement.CheckVerificationEquation(publicNonces, challenge, allResponses));
+
+			// Ensure that verifier rejects invalid transcripts
+			Assert.False(statement.CheckVerificationEquation(publicNonces, new Scalar(17), allResponses));
+			Assert.False(statement.CheckVerificationEquation(simulatedPublicNonces, challenge, allResponses));
+			Assert.False(statement.CheckVerificationEquation(publicNonces, challenge, simulatedResponses));
+		}
+
+		[Fact]
+		public void StatementThrows()
+		{
+			// Jagged equation matrices are not allowed
+			Assert.ThrowsAny<ArgumentException>(() => new Statement(
+				new Equation(GroupElement.Infinity, new GroupElementVector(Generators.Gg, Generators.Gh)),
+				new Equation(GroupElement.Infinity, new GroupElementVector(Generators.G))
+			));
+
+			// Null generators are not allowed
+			Assert.ThrowsAny<ArgumentException>(() => new Statement(
+				new Equation(GroupElement.Infinity, new GroupElementVector(Generators.Gg, Generators.Gh)),
+				new Equation(GroupElement.Infinity, new GroupElementVector(Generators.G, null!))
+			));
+		}
+
+		[Fact]
+		public void KnowledgeThrows()
+		{
+			var x = new Scalar(42);
+			var a = x * Generators.Gg;
+			var b = x * Generators.Gh;
+
+			var statement = new Statement(
+				new Equation(a, new GroupElementVector(Generators.Gg)),
+				new Equation(b, new GroupElementVector(Generators.Gh))
+			);
+
+			// Wrong size for witness
+			Assert.ThrowsAny<ArgumentException>(() => new Knowledge(statement, new ScalarVector(Scalar.Zero, Scalar.Zero)));
+
+			// Incorrect witness
```

Comment:
This comment does not help me to understand it better. Unless I open IDE and start to examine the code, I'm not sure what throws that `ArgumentException`. Incorrect in what way?

Reply:
is this clearer?

```suggestion
			// Incorrect witness (multiplying by generators does not produce the public point)
```

i'm not sure if `ArgumentException` is appropriate, I cargo culted that from nopara's original code, but in the current version Lucas has moved it to `Equation.VerifySolution`)

Reply:
Yes